### PR TITLE
New alternatives to layer norm

### LINKED ISSUE #1113: A couple LayerNorm variants
**Is your feature request related to a current problem? Please describe.**

- LayerNorm with no bias term.

- RMSNorm from "Root mean square layer normalization."

### PR DESCRIPTION
**Pull request by gdevos010**
Fixes #1113.

minor typos.

### CODE REVIEW

**Comment by gdevos010** [comment]
These could be added to N-Beats and N-Hits whenever the layerNorm issue gets worked out

**Comment by gdevos010** [comment]
Not exactly the biggest improvement but here is RMSNorm outperforming LayerNorm on the ice cream example
image
image

**Comment by gdevos010** [comment]
@hrzn Can we retry these tests? All the tests pass locally.

**Comment by dennisbader** [comment]
Hi @gdevos010 and thanks for another PR.
Tests are failing when you upgrade PyTorch Lightning and PyTorch to the newest versions.
I'm working on a fix in #1124. 

Could you have a look at the FeedForward parts that I changed for our TransformerModel? I believe you implemented those so I'd like to have your opinion on it.
The issue was that all our custom FeedForward layers were ignored by the PyTorch torch.nn.Transformer in previous PyTorch versions.

**Comment by gdevos010** [comment]
I also found PowerNorm in my search. 
Forecast_Air Passengers TFT (QR) (LayerNorm)


Forecast_Air Passengers TFT (QR) (PowerNorm)

File: darts/models/components/LayerNormVariants.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import torch
+from torch import nn
+
+
+class ScaleNorm(nn.Module):
```
Comment (hrzn) [inline darts/models/components/LayerNormVariants.py:29]: [Here](https://github.com/tnq177/transformers_without_tears/blob/25026061979916afb193274438f7097945acf9bc/layers.py#L132) `self.scale` is initialized as `Parameter(torch.tensor(scale))`. Also is `self.g` meant to be trainable? In the implementation I linked it's fixed to 1. How confident are you about this implementation?

File: darts/models/components/LayerNormVariants.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+"""
```
Comment (hrzn) [inline darts/models/components/LayerNormVariants.py:1]: Could you name the file `layer_norm_variants.py`?
We don't use CameCase for files.

File: darts/models/components/PowerNorm.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
+# File   : MaskPowerNorm.py
```
Comment (hrzn) [inline darts/models/components/PowerNorm.py:1]: Same here, could you rename to `power_norm.py` ? Actually I'd be inclined to put them all in a single `layer_norm_variants.py` file.

File: darts/models/forecasting/tft_model.py [hunk 4]
```diff
@@ -121,6 +125,18 @@ def __init__(
         self.dropout = dropout
         self.add_relative_index = add_relative_index
 
+        if norm_type == "LayerNorm":
```
Comment (hrzn) [inline darts/models/forecasting/tft_model.py:128]: How about we also optionally accept a `nn.Module` here, and resolve the string only if `isinstance(self.norm_type, str)` ? That would allow users to also specify their own modules.

File: darts/models/forecasting/tft_model.py [hunk 5]
```diff
@@ -121,6 +125,18 @@ def __init__(
...
+            self.layerNorm = nn.LayerNorm
```
Comment (hrzn) [inline darts/models/forecasting/tft_model.py:129]: Could you rename to `self.layer_norm` ? (i.e., snake_case)

File: darts/models/forecasting/tft_model.py [hunk 6]
```diff
@@ -637,6 +668,7 @@ def __init__(
         add_relative_index: bool = False,
         loss_fn: Optional[nn.Module] = None,
         likelihood: Optional[Likelihood] = None,
+        norm_type: str = "LayerNorm",
```
Comment (hrzn) [inline darts/models/forecasting/tft_model.py:671]: how about renaming to `layer_norm_variant` ? To make it clear it touches the layer norm part of the transformer.

File: darts/models/forecasting/tft_submodels.py [hunk 7]
```diff
@@ -273,6 +281,7 @@ def __init__(
         skip_size: int = None,
         trainable_add: bool = False,
         dropout: float = None,
+        layerNorm: nn.Module = nn.LayerNorm,
```
Comment (hrzn) [inline darts/models/forecasting/tft_submodels.py:284]: `layer_norm`

File: darts/models/forecasting/tft_submodels.py [hunk 8]
```diff
@@ -303,6 +315,7 @@ def __init__(
         dropout: float = 0.1,
         context_size: int = None,
         residual: bool = False,
+        layerNorm: nn.Module = nn.LayerNorm,
```
Comment (hrzn) [inline darts/models/forecasting/tft_submodels.py:318]: `layer_norm`

File: darts/models/forecasting/tft_submodels.py [hunk 9]
```diff
@@ -375,6 +390,7 @@ def __init__(
         context_size: int = None,
         single_variable_grns: Optional[Dict[str, _GatedResidualNetwork]] = None,
         prescalers: Optional[Dict[str, nn.Linear]] = None,
+        layerNorm: nn.Module = nn.LayerNorm,
```
Comment (hrzn) [inline darts/models/forecasting/tft_submodels.py:393]: `layer_norm`

File: darts/tests/models/components/test_LayerNormVariants.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+from darts.logging import get_logger
```
Comment (hrzn) [inline darts/tests/models/components/test_LayerNormVariants.py:1]: rename file to snake_case

File: darts/tests/models/components/test_LayerNormVariants.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    logger.warning("Torch not available. Loss tests will be skipped.")
+    TORCH_AVAILABLE = False
+
+
+if TORCH_AVAILABLE:
+    from darts.models.components.LayerNormVariants import RMSNorm, ScaleNorm
+    from darts.models.components.PowerNorm import MaskPowerNorm
+    from darts.tests.base_test_class import DartsBaseTestClass
+
+    class LayerNormVariantsTestCase(DartsBaseTestClass):
```
Comment (hrzn) [inline darts/tests/models/components/test_LayerNormVariants.py:19]: These tests are good, but slightly minimal :) Could you maybe add a test checking correctness of the norm computation with some simple toy values?
Also it'd be good to add a test or two for the TFT, for instance checking that the outputs (on an untrained model) are different for different norm layers.

**Review by hrzn — CHANGES_REQUESTED** [review]
Thanks, nice initiative! I'm pointing out a few things to fix before it can be merged.

File: darts/models/components/LayerNormVariants.py [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+class RMSNorm(nn.Module):
+    """An alternate to layer normalization, without mean centering and the learned bias [1]
+
+    References
+    ----------
+    .. [1] Zhang, Biao, and Rico Sennrich. "Root mean square layer normalization." Advances in Neural Information
+           Processing Systems 32 (2019).
+    """
+
+    def __init__(self, dim, eps=1e-8):
```
[the remark this answers is not in the document]
Comment (gdevos010) [reply darts/models/components/LayerNormVariants.py]: It's from lucidrains' x-transformers repo, so I assumed it was correct, but I see your point. The official implementation also looks different. I will add the official imprmentation and ask lucidrains about it

https://github.com/lucidrains/x-transformers/blob/main/x_transformers/x_transformers.py#L351-L371

Official 
https://github.com/bzhangGo/rmsnorm/blob/master/rmsnorm_torch.py

File: darts/models/components/LayerNormVariants.py (hunk 1 above)
Comment (gdevos010) [reply darts/models/components/LayerNormVariants.py]: Im not sure why lucidrains' implementation is different.

**Comment by gdevos010** [comment]
@hrzn I asked lucidrain about it and here is his [response](https://github.com/lucidrains/x-transformers/issues/101). Let me know what you think and how we want to add it

**Comment by lucidrains** [comment]
yea, so the only two alternatives to layernorm i could recommend is (1) layernorm, but without the bias and (2) rmsnorm

scalenorm i've heard mixed results. i wouldn't risk using that yet

**Comment by gdevos010** [comment]
I removed ScaleNorm as suggested by lucidrains

**Comment by hrzn** [comment]
> yea, so the only two alternatives to layernorm i could recommend is (1) layernorm, but without the bias (removing biases from transformers has reportedly increased stability) and (2) rmsnorm
> 
> scalenorm i've heard mixed results. i wouldn't risk using that yet

Thanks!

**Comment by hrzn** [comment]
@gdevos010 any update on this PR? I think it's not far from being mergeable.

**Comment by gdevos010** [comment]
@hrzn The three variants are now `LayerNorm`, `LayerNormNoBias` and `RMSNorm`. I will post comparison on the sunspot dataset shortly.

**Comment by hrzn** [comment]
> @hrzn The three variants are now `LayerNorm`, `LayerNormNoBias` and `RMSNorm`. I will post comparison on the sunspot dataset shortly.

Great, thanks!

**Comment by gdevos010** [comment]
I definitely went down a rabbit hole trying to make these graphs. Here are 9 examples. 

Backtest_batch_size=128,norm_type=LayerNorm
Backtest_batch_size=128,norm_type=LayerNormNoBias
Backtest_batch_size=128,norm_type=RMSNorm


Backtest_batch_size=256,norm_type=LayerNorm
Backtest_batch_size=256,norm_type=LayerNormNoBias
Backtest_batch_size=256,norm_type=RMSNorm


Backtest_batch_size=512,norm_type=LayerNorm
Backtest_batch_size=512,norm_type=LayerNormNoBias
Backtest_batch_size=512,norm_type=RMSNorm

File: darts/models/forecasting/tft_model.py [hunk 13]
```diff
@@ -706,6 +732,9 @@ def __init__(
         likelihood
             The likelihood model to be used for probabilistic forecasts. By default, the TFT uses
             a ``QuantileRegression`` likelihood.
+        norm_type: str
```
Comment (hrzn) [inline darts/models/forecasting/tft_model.py:735]: Could we optionally allow providing a custom `nn.Module` ? That might be nice to quickly test potential new variants of normalisation layers in the future. WDYT?

**Review by hrzn — COMMENTED** [review]
Thanks @gdevos010 !
There are only a few relatively minor things remaining:
* Could we optionally support `nn.Module` (instead of `str` only) to specify the type of layer norms?
* Could you perhaps also extend this to the `TransformerModel`? If that's too much effort we can leave it out of this PR.
* A few minor other comments

File: darts/models/forecasting/tft_model.py (hunk 13 above)
Comment (gdevos010) [reply darts/models/forecasting/tft_model.py]: Done

**Comment by gdevos010** [comment]
@hrzn I added it to the `TransformerModel`. The implementation is a bit clucky but to support both GLU variants and layer norm individually while maintaining the default behavior was a bit tricky. If you have suggestions on simplifying it, let me know.

File: darts/models/forecasting/transformer_model.py [hunk 14]
```diff
@@ -27,21 +27,35 @@
...
     """Generates an Encoder or Decoder with one of Darts' Feed-forward Network variants.
     Parameters
     ----------
     coder_cls
         Either `torch.nn.TransformerEncoder` or `...TransformerDecoder`
     layer_cls
-        Either `darts.models.components.transformer.CustomFeedForwardEncoderLayer` or
-        `...CustomFeedForwardDecoderLayer`
+        Either `darts.models.components.transformer.CustomFeedForwardEncoderLayer`,
+        `...CustomFeedForwardDecoderLayer`, `nn.TransformerEncoderLayer`, or `nn.TransformerDencoderLayer`.
```
Comment (hrzn) [inline darts/models/forecasting/transformer_model.py:47]:
```suggestion
        `...CustomFeedForwardDecoderLayer`, `nn.TransformerEncoderLayer`, or `nn.TransformerDecoderLayer`.
```

File: darts/tests/models/forecasting/test_transformer_model.py [hunk 15]
```diff
@@ -49,6 +49,7 @@ class TransformerModelTestCase(DartsBaseTestClass):
             dim_feedforward=2048,
             dropout=0.1,
             activation="relu",
+            norm_type=None,
```
Comment (hrzn) [inline darts/tests/models/forecasting/test_transformer_model.py:52]: Was this needed? Isn't this the default value?

File: darts/models/forecasting/tft_model.py [hunk 16]
```diff
@@ -699,13 +728,16 @@ def __init__(
...
-        loss_fn
+        loss_fn: nn.Module
             PyTorch loss function used for training. By default, the TFT model is probabilistic and uses a
             ``likelihood`` instead (``QuantileRegression``). To make the model deterministic, you can set the `
             `likelihood`` to None and give a ``loss_fn`` argument.
         likelihood
             The likelihood model to be used for probabilistic forecasts. By default, the TFT uses
             a ``QuantileRegression`` likelihood.
+        norm_type: str | nn.Module
+            The type of LayerNorm variant to use.  Default: ``LayerNorm``. Avaliable options are
```
Comment (hrzn) [inline darts/models/forecasting/tft_model.py:739]:
```suggestion
            The type of LayerNorm variant to use.  Default: ``LayerNorm``. Available options are
```

**Review by hrzn — COMMENTED** [review]
Thanks for this iteration. It starts looking good. I would feel a bit better if before merging we could have another couple of unit tests that build & run TransformerModel and TFTModel with at least one non-default norm layer...

**Review by hrzn — APPROVED** [review]
Thanks @gdevos010 !
LGTM. Are we good to merge?

**Comment by gdevos010** [comment]
@hrzn sorry I took so long adding the tests. We are still trying to move :(

**Review by hrzn — APPROVED** [review]

**Comment by hrzn** [comment]
> @hrzn We are good to merge! Sorry I took so long adding the tests. We are still trying to move :(

Looks great! Merging now. Thanks!
